Quickstart guide to manage vantage cloud compute cluster thought airflow (#210)
Teradata/quickstarts: modules/vantagecloud-lake/pages/vantagecloud-lake-compute-cluster-airflow.adoc

> = Manage Vantage Cloud Lake Compute Cluster with Airflow

Reviewer: ```suggestion
= Manage Vantage Cloud Lake Compute Clusters with Apache Airflow
```

:experimental:
:page-author: Satish Chinthanippu
:page-email: satish.chinthanippu@teradata.com
:page-revdate: July 11th, 2024

> :description: Manage Vantage Cloud Lake Compute Cluster with Airflow

Reviewer: ```suggestion
:description: Manage Vantage Cloud Lake Compute Clusters with Airflow
```

:keywords: data warehouses, compute storage separation, teradata, vantage, cloud data platform, business intelligence, enterprise analytics, airflow, workflow, teradatasql, ipython-sql, cloud computing, machine learning, vantagecloud, vantagecloud lake, lake
:dir: vantagecloud-lake-compute-cluster-airflow
:auxdir: vantagecloud-lake-compute-cluster-airflow

== Overview

> This tutorial demonstrates the process of utilizing the https://airflow.apache.org/docs/apache-airflow-providers-teradata/stable/operators/index.html[Teradata Airflow Compute Cluster Operators] to manage the Vantage Cloud Lake Compute Cluster. The objective is to execute dbt models on the Lake Compute Cluster.

Reviewer: ```suggestion
This tutorial demonstrates the process of utilizing the https://airflow.apache.org/docs/apache-airflow-providers-teradata/stable/operators/index.html[Teradata Airflow Compute Cluster Operators] to manage Vantage Cloud Lake Compute Clusters. The objective is to execute dbt models on the VantageCloud Lake Compute Clusters.
```

== Prerequisites

> * Access to a VantageCloud Lake environment

Reviewer 2: Should we say "Ensure you have the necessary credentials and access rights to use Teradata Vantage Lake."

Author: Modified as Ensure you have the necessary credentials and access rights to use Teradata Vantage Lake.

----
export AIRFLOW_HOME=~/airflow
----
2. Install `apache-airflow` stable version 2.9.2 from PyPI repository.:
+
[tabs]
====
Windows::
+
--
[source, bash]
----
AIRFLOW_VERSION=2.9.2
PYTHON_VERSION="$(python --version | cut -d " " -f 2 | cut -d "." -f 1-2)"
CONSTRAINT_URL="https://raw.githubusercontent.com/apache/airflow/constraints-${AIRFLOW_VERSION}/constraints-${PYTHON_VERSION}.txt"
pip install "apache-airflow==${AIRFLOW_VERSION}" --constraint "${CONSTRAINT_URL}"
----
--
MacOS::
+
--
[source, bash]
----
AIRFLOW_VERSION=2.9.2
PYTHON_VERSION="$(python3 --version | cut -d " " -f 2 | cut -d "." -f 1-2)"
CONSTRAINT_URL="https://raw.githubusercontent.com/apache/airflow/constraints-${AIRFLOW_VERSION}/constraints-${PYTHON_VERSION}.txt"
pip install "apache-airflow==${AIRFLOW_VERSION}" --constraint "${CONSTRAINT_URL}"
----
--
Linux::
+
--
[source, bash]
----
AIRFLOW_VERSION=2.9.2
PYTHON_VERSION="$(python3 --version | cut -d " " -f 2 | cut -d "." -f 1-2)"
CONSTRAINT_URL="https://raw.githubusercontent.com/apache/airflow/constraints-${AIRFLOW_VERSION}/constraints-${PYTHON_VERSION}.txt"
pip install "apache-airflow==${AIRFLOW_VERSION}" --constraint "${CONSTRAINT_URL}"
----
--
====

> 3. Install the Airflow Teradata provider stable version from PyPI repository.

Reviewer: ```suggestion
3. Install the Apache Airflow Teradata provider stable version from PyPI repository.
```

+
[source, bash]
----
pip install "apache-airflow-providers-teradata"
----

== Install dbt
1. Create a new python environment to manage dbt and its dependencies. Activate the environment:
+
[tabs]
====
Windows::
+
--
[source, bash]
----
python -m venv env
source env/bin/activate
----
--
MacOS::
+
--
[source, bash]
----
python3 -m venv env
source env/bin/activate
----
--
Linux::
+
--
[source, bash]
----
python3 -m venv env
source env/bin/activate
----
--
====

2. Install `dbt-teradata` and `dbt-core` modules:
+
[source, bash]
----
pip install dbt-teradata dbt-core
----

> == Create a database

Reviewer: @sc250072 I feel maybe we can mention that these SQL commands should be executed from a database client, either lake editor, or dbeaver, something like that.

Author: Added as note about this

----

== Create a database user

Let's create `lake_user` user in Vantage lake instance.

[source, teradata-sql]
----
CREATE USER lake_user
AS PERMANENT = 1000000,
PASSWORD = lake_user,
SPOOL = 1200000,
DEFAULT DATABASE = jaffle_shop;
----

== Grant access to user

Let's provide required privileges to user `lake_user` to manage compute cluster.

[source, teradata-sql]
----
GRANT CREATE COMPUTE GROUP To lake_user;
GRANT DROP COMPUTE GROUP TO lake_user;
GRANT SELECT ON DBC TO lake_user;
GRANT ALL ON jaffle_shop TO lake_user;
----

== Setup dbt project

1. Clone the jaffle_shop repository and cd into the project directory:
+
[source, bash]
----
git clone https://github.com/Teradata/jaffle_shop-dev.git jaffle_shop
cd jaffle_shop
----
2. Create `profiles.yml` file. Add the following config to `profile.yml` file. Adjust `<host>` to match your Teradata Vantage lake instance.
+
[source, yaml]
----
jaffle_shop:
  outputs:
    dev:
      type: teradata
      host: <host>
      user: lake_user
      password: lake_user
      logmech: TD2
      schema: jaffle_shop
      tmode: ANSI
      threads: 1
      timeout_seconds: 300
      priority: interactive
      retries: 1
  target: dev
----

> == Define DAG in Airflow

Reviewer: @sc250072 Maybe we can add a bit more details on how to name the file and where to store it.

Author: Yes. Added with details on how and where to save

> == Define DAG in Airflow

Reviewer: ```suggestion
== Define DAG in Apache Airflow
```

    )
    # suspend compute cluster
    compute_cluster_suspend_operation = TeradataComputeClusterSuspendOperator(
        task_id="compute_cluster_suspend_operation",
        compute_profile_name="dbt_transformation_profile",
        compute_group_name="dbt_transformation_group",
    )
    # decommission compute cluster
    compute_cluster_decommission_operation = TeradataComputeClusterDecommissionOperator(
        task_id="compute_cluster_decommission_operation",
        compute_profile_name="dbt_transformation_profile",
        compute_group_name="dbt_transformation_group",
        delete_compute_group=True,
    )
    # remove compute group for user
    remove_compute_group_from_user = TeradataOperator(
        task_id="remove_compute_group_from_user",
        sql=r"""
        MODIFY USER lake_user
        AS COMPUTE GROUP = NULL
        """,
    )

    (
        compute_cluster_provision_operation
        >> assign_compute_group_user
        >> compute_cluster_resume_operation
        >> transform_data
        >> compute_cluster_suspend_operation
        >> compute_cluster_decommission_operation
        >> remove_compute_group_from_user
    )
----

> == Configure Airflow

Reviewer: ```suggestion
== Configure Apache Airflow
```

1. Configure the listed environment variables to activate the test connection button, preventing the loading of sample DAGs and default connections in Airflow UI.
+
[source, bash]
  export AIRFLOW__CORE__TEST_CONNECTION=Enabled
  export AIRFLOW__CORE__LOAD_EXAMPLES=false
  export AIRFLOW__CORE_LOAD_DEFAULT_CONNECTIONS=false

2. Define the path of jaffle_shop project as an environment variable `dbt_project_home_dir`.
+
[source, bash]
export dbt_project_home_dir=../../jaffle_shop
+
NOTE: Change `/../../` to path of jaffle_shop project path.

3. Define the virtual environment path where dbt-teradata installed in <<Install dbt>> as an environment variable `dbt_venv_dir`.
[source, bash]
set dbt_venv_dir=/../../dbt_env/bin/activate
+
NOTE: Change `/../../` to path where virtual environment path defined.

> == Start Airflow web server

Reviewer: ```suggestion
== Start Apache Airflow web server
```

1. Run airflow web server
+
[source, bash]
----
airflow standalone
----
5. Access the airflow UI. Visit https://localhost:8080 in the browser and log in with the admin account details shown in the terminal.

> == Define Airflow connection to Vantage Cloud Lake

Reviewer: ```suggestion
== Define Apache Airflow connection to Vantage Cloud Lake
```

Reviewer: Added a couple of more editions basically regarding Apache Product guidelines, specifically that all titles and subtitles should include the full name of the project.

Author: Included Apache at specified files

1. Click on Admin - Connections
2. Click on + to define new connection to Teradata vantage cloud lake instance.
3. Define new connection with id `teradata_lake` with Teradata vantage cloud lake instance details.
* Connection Id: teradata_lake
* Connection Type: Teradata.
* Database Server URL (required): Teradata vantage cloud lake instance hostname to connect to.
* Database: jaffle_shop
* Login (required): lake_user
* Password (required): lake_user

== Load DAG

> Copy this dag file to $AIRFLOW_HOME/files/dags. Airflow loads DAGs from Python source files, which it looks for inside its configured DAG_FOLDER - $AIRFLOW_HOME/files/dags directory.

Reviewer: Is this the DAG we created above?

----
--
Linux::
+
--
[source, bash]
----
AIRFLOW_VERSION=2.9.2
PYTHON_VERSION="$(python3 --version | cut -d " " -f 2 | cut -d "." -f 1-2)"
CONSTRAINT_URL="https://raw.githubusercontent.com/apache/airflow/constraints-${AIRFLOW_VERSION}/constraints-${PYTHON_VERSION}.txt"
pip install "apache-airflow==${AIRFLOW_VERSION}" --constraint "${CONSTRAINT_URL}"
----
--
====

3. Install the Apache Airflow Teradata provider stable version from PyPI repository.
+
[source, bash]
----
pip install "apache-airflow-providers-teradata"
----

== Install dbt
1. Create a new python environment to manage dbt and its dependencies. Activate the environment:
+
[tabs]
====
Windows::
+
--
[source, bash]
----
python -m venv env
source env/bin/activate
----
--
MacOS::
+
--
[source, bash]
----
python3 -m venv env
source env/bin/activate
----
--
Linux::
+
--
[source, bash]
----
python3 -m venv env
source env/bin/activate
----
--
====

2. Install `dbt-teradata` and `dbt-core` modules:
+
[source, bash]
----
pip install dbt-teradata dbt-core
----

== Create a database

NOTE: Use a database client like BTEQ, Lake Editor, Teradata Studio or  dbeaver to execute `CREATE DATABASE` query.

> Let's create `jaffle_shop` database in Vantage lake instance with TD_OFSSTORAGE as default.

Reviewer: ```suggestion
Let's create the `jaffle_shop` database in the VantageCloud Lake instance with TD_OFSSTORAGE as default.
```

[source, teradata-sql]
----
CREATE DATABASE jaffle_shop
AS DEFAULT STORAGE = TD_OFSSTORAGE OVERRIDE ON ERROR,
PERMANENT = 120e6, -- 120MB
    SPOOL = 120e6; -- 120MB
----

== Create a database user

> NOTE: Use a database client like BTEQ, Lake Editor, Teradata Studio or  dbeaver to execute `CREATE USER` query.

Reviewer: The note above, might be enough I think we can eliminate this one.

> Let's create `lake_user` user in Vantage lake instance.

Reviewer: ```suggestion
Let's create a `lake_user` user in the VantageCloud Lake instance.
```

[source, teradata-sql]
----
CREATE USER lake_user
AS PERMANENT = 1000000,
PASSWORD = lake_user,
SPOOL = 1200000,
DEFAULT DATABASE = jaffle_shop;
----

== Grant access to user

> NOTE: Use a database client like BTEQ, Lake Editor, Teradata Studio or  dbeaver to execute `GRANT ACCESS` queries.

Reviewer: Same as above, this line can be eliminated

> Let's provide required privileges to user `lake_user` to manage compute cluster.

Reviewer: ```suggestion
Let's provide the required privileges to the user `lake_user` to manage compute clusters.
```

[source, teradata-sql]
----
GRANT CREATE COMPUTE GROUP To lake_user;
GRANT DROP COMPUTE GROUP TO lake_user;
GRANT SELECT ON DBC TO lake_user;
GRANT ALL ON jaffle_shop TO lake_user;
----

== Setup dbt project

> 1. Clone the jaffle_shop repository and cd into the project directory:

Reviewer: ```suggestion
1. Clone the jaffle_shop dbt demo repository and cd into the project directory:
```

+
[source, bash]
----
git clone https://github.com/Teradata/jaffle_shop-dev.git jaffle_shop
cd jaffle_shop
----

> 2. Create `profiles.yml` file. Add the following config to `profile.yml` file. Adjust `<host>` to match your Teradata Vantage lake instance.

Reviewer: ```suggestion
2. Create `profiles.yml` file. Add the following config to `profile.yml` file. Adjust `<host>` to match your Teradata VantageCloud Lake instance.
```

+
[source, yaml]
----
jaffle_shop:
  outputs:
    dev:
      type: teradata
      host: <host>
      user: lake_user
      password: lake_user
      logmech: TD2
      schema: jaffle_shop
      tmode: ANSI
      threads: 1
      timeout_seconds: 300
      priority: interactive
      retries: 1
  target: dev
----

== Configure Apache Airflow

1. Configure the listed environment variables to activate the test connection button, preventing the loading of sample DAGs and default connections in Airflow UI.
+
[source, bash]
  export AIRFLOW__CORE__TEST_CONNECTION=Enabled
  export AIRFLOW__CORE__LOAD_EXAMPLES=false
  export AIRFLOW__CORE_LOAD_DEFAULT_CONNECTIONS=false

2. Define the path of jaffle_shop project as an environment variable `dbt_project_home_dir`.
+
[source, bash]
export dbt_project_home_dir=../../jaffle_shop
+

> NOTE: Change `/../../` to path of jaffle_shop project path.

Reviewer: ```suggestion
NOTE: You might need to change `/../../` to the specific path where you cloned the jaffle_shop project.
```

> 3. Define the virtual environment path where dbt-teradata installed in <<Install dbt>> as an environment variable `dbt_venv_dir`.

Reviewer: ```suggestion
3. Define the path to the virtual environment where dbt-teradata was installed as an environment variable `dbt_venv_dir`.
```

[source, bash]
set dbt_venv_dir=/../../dbt_env/bin/activate
+

> NOTE: Change `/../../` to path where virtual environment path defined.

Reviewer: ```suggestion
NOTE: You might need to change `/../../` to the specific path where the virtual environment is located.
```

== Start Apache Airflow web server
1. Run airflow web server
+
[source, bash]
----
airflow standalone
----

> 2. Access the airflow UI. Visit https://localhost:8080 in the browser and log in with the admin account details shown in the terminal.

Reviewer: Let's add the image of the console with the credentials here.

> == Define Apache Airflow connection to Vantage Cloud Lake

Reviewer: ```suggestion
== Define a connection to VantageCloud Lake in Apache Airflow
```

1. Click on Admin - Connections

> 2. Click on + to define new connection to Teradata vantage cloud lake instance.

Reviewer: ```suggestion
2. Click on + to define new connection to the Teradata VantageCloud Lake instance.
```

> 3. Define new connection with id `teradata_lake` with Teradata vantage cloud lake instance details.

Reviewer: ```suggestion
3. Define a new connection with the Teradata VantageCloud Lake instance details. Assign the id `teradata_lake`
```

* Connection Id: teradata_lake
* Connection Type: Teradata.

> * Database Server URL (required): Teradata vantage cloud lake instance hostname to connect to.

Reviewer: ```suggestion
* Database Server URL (required): Teradata VantageCloud Lake instance hostname or IP to connect to.
```

* Database: jaffle_shop
* Login (required): lake_user
* Password (required): lake_user

== Define DAG in Apache Airflow

> Dags in airflow will be defined as python file. Similarly, define below dag to run dbt transformations defined in `jaffle_shop` dbt project using vantage cloud lake compute cluster. Copy below python code and save it as `airflow-vcl-compute-clusters-manage.py` under directory $AIRFLOW_HOME/files/dags.

Reviewer: ```suggestion
Dags in airflow are defined as python files. The dag below runs the dbt transformations defined in the `jaffle_shop` dbt project using VantageCloud Lake compute clusters. Copy the python code below and save it as `airflow-vcl-compute-clusters-manage.py` under the directory $AIRFLOW_HOME/files/dags.
```

    )
    # suspend compute cluster
    compute_cluster_suspend_operation = TeradataComputeClusterSuspendOperator(
        task_id="compute_cluster_suspend_operation",
        compute_profile_name="dbt_transformation_profile",
        compute_group_name="dbt_transformation_group",
    )
    # decommission compute cluster
    compute_cluster_decommission_operation = TeradataComputeClusterDecommissionOperator(
        task_id="compute_cluster_decommission_operation",
        compute_profile_name="dbt_transformation_profile",
        compute_group_name="dbt_transformation_group",
        delete_compute_group=True,
    )
    # remove compute group for user
    remove_compute_group_from_user = TeradataOperator(
        task_id="remove_compute_group_from_user",
        sql=r"""
        MODIFY USER lake_user
        AS COMPUTE GROUP = NULL
        """,
    )

    (
        compute_cluster_provision_operation
        >> assign_compute_group_user
        >> compute_cluster_resume_operation
        >> transform_data
        >> compute_cluster_suspend_operation
        >> compute_cluster_decommission_operation
        >> remove_compute_group_from_user
    )
----
== Load DAG

> When the dag file is copied to $AIRFLOW_HOME/files/dags, apache airflow loads the dag to airflow UI.

Reviewer: ```suggestion
Once the dag file is copied to $AIRFLOW_HOME/files/dags, apache airflow loads the dag to the airflow UI.
```

== Run DAG

Run the dag as shown in below image.

image::{dir}/airflow-dag-run.png[Run dag,align="left" width=75%]

== Summary

> In this quick start guide, we explored how to utilize Teradata Vantage Lake Compute Cluster to execute dbt transformations using Airflow compute cluster operators.

Reviewer: ```suggestion
In this quick start guide, we explored how to utilize Teradata VantageCloud Lake compute clusters to execute dbt transformations using Teradata compute cluster operators for Airflow.
```

> Run the dag as shown in below image.

Reviewer: ```suggestion
Run the dag as shown in the image below.
```
